portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,19 +14,52 @@
 
 package com.google.sps.servlets;
 
+import com.google.gson.Gson;
 import java.io.IOException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import java.util.List; 
+import java.util.ArrayList;
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
+  private List<String> messages;
+
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello Thomas!</h1>");
+    
+    String json = convertToJsonUsingGson(messages);
+    response.setContentType("application/json;");
+    response.getWriter().println(json);
   }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException{
+
+      String message = getMessage(request);
+
+      if (messages == null){
+          messages = new ArrayList<>();
```

Comment:
You might consider defining an init function to initialize things like this. e.g. https://github.com/googleinterns/step/blob/0218d07ca16ad8ead819a051f69edb563d030f86/walkthroughs/week-3-server/examples/random-quotes/src/main/java/com/google/sps/servlets/RandomQuoteServlet.java#L32

Reply:
Agreed, putting this in `init` is better because you should have `messages != null` as a class invariant -- i.e., something which is always true about the class. Then, in your public methods like `doPost` and `doGet` you won't need to check whether messages is null.

> The class invariant constrains the state stored in the object. Class invariants are established during construction and constantly maintained between calls to public methods.

https://en.wikipedia.org/wiki/Class_invariant

In the case of HttpServlets, init() is preferred (as opposed to the constructor) because the server can reuse old servlet objects to avoid the overhead of constructing new ones. Then init() is responsible for setting up the class's initial state/invariants.